---
title: "BCB420 Assignment 1"
author: "Kevin Zhu"
output:
  html_document:
    df_print: paged
    doc: yes
    toc: true
    theme: united
  html_notebook: null
---

## Objective of Assignment 1
The objective of this assignment is to produce a clean, normalized dataset that
will be used for the following tasks in this course. 

## Dataset Overview
γ-Secretase is a protease known to cleave the CTFs of amyloid-beta precursor 
protein (APP) at its gamma site. It is also known to cleave other substrates, 
and has been associated pathways governing cell differentiation. However, we 
still do not have a full understanding of what γ-Secretase does to 
non-proliferating cells (such as neurons). 

Dysfunction of lipid metabolism neurons has been linked with 
Alzheimer's Disease. In mouse embryonic fibroblasts (MEFs), γ-secretase 
inhibition lowered cellular cholesterol ester levels and reduced endocytosis of 
the low-density lipoprotein receptor (LDLR). However, No studies were carried 
out in human neurons, whose cholesterol metabolism is uniquely different from 
that of glia or peripheral cells.

The authors conducted RNA-seq experiments six biological replicates of induced 
human neuron cells. There were three treatment protocols: Control, DAPT, and 
LY411575. Specifically, DAPT and LY411575 treatments pharmaceutically induced 
chronic inhibition of γ-secretase. The associated GEO ID for this dataset is 
**GSE206102**. 

## Installing dependencies and dowloading the packages
<br>
These dependencies has are already included with the command in the docker file

```{r, message=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")}

if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  BiocManager::install("GEOmetadb")}

if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  install.packages("knitr")}

if (!requireNamespace("edgeR", quietly = TRUE)){
  BiocManager::install("edgeR")}

if (!requireNamespace("biomaRt", quietly = TRUE)){
  BiocManager::install("biomaRt")}

if (!requireNamespace("DBI", quietly = TRUE)){
  install.packages("DBI")}

if (!requireNamespace("GEOquery", quietly = TRUE)){
  BiocManager::install("GEOquery")}

library(GEOquery)
library(knitr)
library(edgeR)
library(biomaRt)
library(dplyr)
library(purrr)
```
## Step 1: Loading the Expression Dataset
This step retrieves the gene expression data from the GEO, using the accession
number **GSE206102**. 

### Introduction of Dataset
```{r, message=FALSE}
dataset_geo_id <- 'GSE206102'

# downloading GEO dataset
gse <- getGEO(dataset_geo_id ,GSEMatrix = FALSE)

# Information about Platform
current_gpl <- names(GPLList(gse))[1]
current_gpl_info <- Meta(getGEO(current_gpl))

list_of_samples <- gse@gsms

```

**GEO ID:** `r dataset_geo_id`

**Dataset Name:** Neuronal γ-secretase regulates lipid metabolism, linking 
cholesterol to synaptic dysfunction in Alzheimer’s disease

**Platform Title:** `r current_gpl_info$title`

**Technology:** `r current_gpl_info$technology`

**Submission Date:** `r current_gpl_info$submission_date`

**Last Update Date:** `r current_gpl_info$last_update_date`

**Organism:** `r current_gpl_info$organism`

```{r, message=FALSE}
# getting the sample descriptions
samples_type <- do.call(rbind, lapply(list_of_samples,
                                      FUN=function(x){
                                        c(x@header$title,
                                          x@header$characteristics_ch1)
                                        }
                                      )
                        )

colnames(samples_type) <- c("title", 
                            "tissue",
                            "cell line", 
                            "cell type",
                            "genotype", 
                            "treatment")

samples_type[,'tissue'] <- gsub(samples_type[,'tissue'], 
                                pattern = "tissue: ", replacement = "")
samples_type[,'cell line'] <- gsub(samples_type[,'cell line'], 
                                   pattern = "cell line: ", replacement = "")
samples_type[,'cell type'] <- gsub(samples_type[,'cell type'], 
                                   pattern = "cell type: ", replacement = "")
samples_type[,'genotype'] <- gsub(samples_type[,'genotype'], 
                                   pattern = "genotype: ", replacement = "")
samples_type[,'treatment'] <- gsub(samples_type[,'treatment'], 
                                   pattern = "treatment: ", replacement = "")

kable(samples_type[1:6,], format = "html")

```

### Downloading the Dataset
Note that an earlier attempt to process the file involved using the supplemental 
file provided on GEO. However, this data was at the transcript-level. To
circumvent this issue, I am using GEO's new beta feature to download gene-level 
raw RNA-seq counts, per Prof. Isserlin's announcement. 

```{r, message=FALSE}

urld <- "https://www.ncbi.nlm.nih.gov/geo/download/?format=file&type=rnaseq_counts"
path <- paste(urld, "acc=GSE206102", 
              "file=GSE206102_raw_counts_GRCh38.p13_NCBI.tsv.gz", sep="&")
tbl <- as.matrix(data.table::fread(path, header=T, colClasses="integer"), 
                 rownames=1)
head(tbl)
```

Now that we have the raw gene counts, we can rename the columns to make them 
easier to understand. We can also see that there are 12 columns instead of 18 
columns. It seems like a few of the samples were omitted from the original 
RNA-seq dataset; we will need to modify the `samples_type` dataframe.

```{r eval=TRUE}
to_keep <- vector()
for (col_name in colnames(tbl)) {
  idx <- grep(col_name, rownames(samples_type))
  to_keep <- c(to_keep, idx)
  new_col_name <- toString(samples_type[, "title"][idx])
  colnames(tbl)[colnames(tbl) == col_name] <- new_col_name
}

samples_type <- samples_type[to_keep, ]

head(tbl)
```

## Step 2: Cleaning the Data
In this step, we will be removing genes with low counts as wll as mapping the
row IDs to the proper HUGO symbols

### Filtering Rows with Low Counts
```{r eval=TRUE}
# Removing low counts; we can play around with the min_samples value
min_samples <- 4
keep <- rowSums( edgeR::cpm(tbl) > 1 ) >= min_samples
filtered_tbl <- tbl[keep, ]
dim(filtered_tbl) 
```

### Converting to HGNC symbols

```{r eval=TRUE}
ncbi_gene_ids <- rownames(filtered_tbl)
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
out_label <- "hgnc_symbol"
hgnc_symbols <- getBM(
  attributes = c("entrezgene_id", "hgnc_symbol"),
  filters = "entrezgene_id",
  values = ncbi_gene_ids,
  mart = ensembl,
  uniqueRows = TRUE
)

# remove duplicates such that each entrezgene_id has a unique HGNC symbol
unique_hgnc_symbols <- group_by(hgnc_symbols, hgnc_symbol) %>% 
  slice(1) %>%
  group_by(entrezgene_id) %>%
  slice(1)

unique_hgnc_symbols <- unique_hgnc_symbols[
  !(unique_hgnc_symbols$hgnc_symbol == ""), ]

is_entrez_unique <- 
  length(unique(unique_hgnc_symbols$entrezgene_id)) == nrow(unique_hgnc_symbols)
is_hgnc_unique <- 
  length(unique(unique_hgnc_symbols$hgnc_symbol)) == nrow(unique_hgnc_symbols)

# Check if both conditions are satisfied
if (is_entrez_unique && is_hgnc_unique) {
  print("Both columns have unique values.")
} else {
  print("Failed")
}

# Merge the table with HGNC symbols
hgnc_tbl <- merge(filtered_tbl, unique_hgnc_symbols, 
                  by.x = "row.names", by.y = "entrezgene_id", all=T)

missing_values <- rownames(hgnc_tbl[is.na(hgnc_tbl$hgnc_symbol), ])
length(missing_values)
```

Unfortunately, there were NCBI IDs in which `biomaRt` was unable to find the 
corresponding HUGO gene symbol. It seems that the majority of them are either 
outdated and thus no longer in the current gene set, or they are pseudogenes 
that do not encode functional proteins. Furthermore, each Entrez ID sometimes
corresponds to multiple HGNC symbols, which means more some duplicates had to 
be removed from the dataset in order to have unique row names. 

```{r eval=TRUE}
# remove rows that do not have HUGO symbol
hgnc_tbl <- hgnc_tbl[!(is.na(hgnc_tbl$hgnc_symbol)), ]

# Set rownames using HGNC symbols and remove extra columns
rownames(hgnc_tbl) <- hgnc_tbl$hgnc_symbol
filtered_tbl <- hgnc_tbl[, -c(1, ncol(hgnc_tbl))]

# Print the updated table
head(filtered_tbl)
```

## Step 3: Normalizing the Data
Now that we have cleaned our data, we can now proceed to the normalization step. 
Before this, we must look at how our data looks pre-normalized. 

### Initial Plotting
```{r eval=TRUE}
# boxplot for pre-normalized dataset
# we do add a pseudocount of 0.25 to ensure negative log values do not explode
data2plot <- log2(cpm(filtered_tbl)+0.25)

boxplot(data2plot,xlab = "Samples", ylab = "log2 CPM",
 las = 2, cex = 0.5, cex.lab = 0.5,
 cex.axis = 0.5, main = "Pre-normalized RNASeq data")

#draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)),
 col = "green", lwd = 0.6, lty = "dashed")
```

Interestingly, despite having downloaded the raw RNA-seq data, the 
overall distributions between the samples look almost identical, 
with medians and quartiles at the same values. We will nevertheless still 
apply a normalization to see what it looks like. This is confirmed by the
density plot, as shown below:

```{r eval=TRUE}
# note that this plotting code is based on Lecture 5 notes

counts_density <- apply(data2plot, 2, density)

# calculate the limits across all the samples
xlim <- 0
ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x))
  ylim <- range(c(ylim, counts_density[[i]]$y))
}

cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))

# plot the first density plot to initialize the plot
plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n",
     ylab="Smoothing density of log2-CPM",
     main="", cex.lab = 0.85)

# plot each line
for (i in 1:length(counts_density))
  lines(counts_density[[i]], col=cols[i], lty=ltys[i])

# create legend
legend("topright", colnames(data2plot),
       col=cols, lty=ltys, cex=0.7,
       border ="blue", text.col = "green4",
       merge = TRUE, bg = "gray90")

```

### Normalization using TMM
```{r eval=TRUE}
filtered_data_matrix <- as.matrix(filtered_tbl)

# groups <- colnames(filtered_tbl)
groups <- samples_type[, "treatment"]
# d = DGEList(counts=filtered_data_matrix, group=groups)
d = DGEList(counts=filtered_tbl, group=groups)
# normalization
d = calcNormFactors(d)
normalized_counts <- cpm(d)
nrow(normalized_counts)
```
```{r eval=TRUE}
# we do add a pseudocount of 0.25 to ensure negative log values do not explode
data2plot_after <- log2(normalized_counts + 0.25)
boxplot(data2plot_after, xlab = "Samples", ylab = "log2 CPM",
 las = 2, cex = 0.5, cex.lab = 0.5,
 cex.axis = 0.5, main = "Normalized RNASeq Samples")
#draw the median on each box plot
abline(h = median(apply(data2plot_after, 2, median)),
 col = "green", lwd = 0.6, lty = "dashed")
```

```{r eval=TRUE}
# note that this plotting code is based on Lecture 5 notes

counts_density <- apply(data2plot_after, 2, density)

# Calculate the limits across all the samples
xlim <- 0
ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x))
  ylim <- range(c(ylim, counts_density[[i]]$y))
}

cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))

# Plot the first density plot to initialize the plot
plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n",
     ylab="Smoothing density of log2-CPM", main="", cex.lab = 0.85)

# Plot each line
for (i in 1:length(counts_density))
  lines(counts_density[[i]], col=cols[i], lty=ltys[i])

# Create legend
legend("topright", colnames(data2plot),
       col=cols, lty=ltys, cex=0.75,
       border ="blue", text.col = "green4",
       merge = TRUE, bg = "gray90")

```

We can see that these results are essentially identical to the plot we made 
earlier. In fact, the normalization factors are all very close to 1, which 
probably means that the data had already been normalized. Note that the dataset 
used in this analysis is supposed to be the raw RNA-seq counts. Thus, the 
original data submitted to GEO had likely been pre-normalized, and NCBI simply
performed the processing work to map the data to gene-level counts.

## Step 4: Initial Analysis and Discussion
In this step, we can perform further analyses to see how the data is 
distributed, as well as how seeing how samples differ across treatments.

### MDS Plotting

We can visualize the sample separation using a multidimenstional scaling plot
post normalization. We can see from the MDS plot below that there does seem to 
be some separation between the control samples and the ones with the 
γ-secretase inhibitors.
```{r eval=TRUE}
plotMDS(d, labels=NULL,pch=1,
  col = c("darkgreen","blue", "red")[factor(samples_type[, "treatment"])]
)
par(xpd = T)
legend("topright",
legend=levels(factor(samples_type[, "treatment"] )),
pch=c(1), col= c("darkgreen","blue", "red"),
cex = 0.7, inset=c(0,-0.23))
```

### Dispersion Plotting
The dispersion analysis is performed below. The plot seems to indicate that the
dispersion is roughly constant across the genes. This may suggest consistent 
technical variability.

```{r eval=TRUE}
model_design <- model.matrix(~samples_type[, "treatment"])
d <- estimateDisp(d, model_design)
plotBCV(d, col.tagwise = "black", col.common = "red", )
```

### Mean Variance Plotting
As shown below, it seems like the data follows a negative binomial distribution, 
with the points following relatively closely the blue line. 

```{r eval=TRUE}
plotMeanVar(d, show.raw.vars = TRUE, show.tagwise.vars = TRUE, NBline = TRUE, 
            show.ave.raw.vars = TRUE, show.binned.common.disp.vars = TRUE, 
            ylim = c(1,1e10))

```

### Writing Out the Data
```{r eval=TRUE}
write.table(normalized_counts, 
            file.path(
              getwd(),
              paste(dataset_geo_id, 
                    "normalized_filtered_RSEM_counts.txt",sep="_")
              ),
            quote = FALSE,sep="\t",row.names = TRUE)
```

### Questions and Answers
**What are the control and test conditions of the dataset?**\
The RNA-seq experiment was conducted over six biological replicates of induced
neuronal cells. Tehre were three treatments: control (0.6% DMSO), DAPT(40 uM), 
and LY411575(2.5 uM). The latter two treatments were used to chronically inhibit
γ-secretase.

**Why is the dataset of interest to you?**\
Alzheimer's disease (AD); it is the most common cause of dementia but there 
haven't been any successful therapeutic interventions for this issue. More 
recently, there have been interesting observations regarding the relationship
between dysfunction in lipid metabolism and neuroinflammation in AD. This 
dataset and its associate paper aims to elucidate this relationship by 
investigating the direct causal relationship between γ-secretase, cholesterol, 
and synapses.

**Were there expression values that were not unique for specific genes?**\
This dataset was originally published at the transcript level using ensembl 
transcript IDs. As such, I used the the NCBI-generated RNA-seq count data, which
had gene-level expression, mapped to unique Entrez Gene IDs. However, when 
mapping to HUGO symbols, there was an issue with `biomaRt` returning multiple
HGNC IDs for a single Entrez Gene ID, resulting in duplicate genes that had to
be removed.

**Were there expression values that couldn't be mapped to HUGO symbols?**\
There were a few genes in the original dataset that likely had outdated 
Entrez Gene IDs, as `biomaRt` was unable to locate their HUGO symbol. After this 
conversion step from transcript to gene-level expression, 
`r length(missing_values)` genes were omitted from the dataset. \

**How many outliers were removed?**\
I did not remove any sample outliers. As shown by the density and box plots, it
seems as though the dataset had already been normalized, as there were no
indication of outliers. Hence, no removal of data was performed in this step. 
The only data that was removed were for low counts; genes were considered 
expressed when it had >1 count in at least `r min_samples` samples.

**How did you handle replicates?**\
The dataset contained six biological replicates, each with three treatments.
However, the data NCBI-generated RNA-seq count data only has twelve total 
samples, with Replicates 1-3 each missing two conditions. Nevertheless, because
the samples did not seem to be highly paired between replicates, the analysis
was performed on the 12 samples. Thus, the final output data table has 12 
columns. Note that when removing genes with low sample counts, I removed entire
rows to keep the data consistent.

**What is the final coverage of your dataset?**\
The final dataset coverage is `r dim(normalized_counts)[1]` genes. There are 
12 columns (each condition has four total replicates).

## References
\setlength{\parindent}{-0.2in}
\setlength{\leftskip}{0.2in}
\setlength{\parskip}{8pt}
\noindent

<div style="text-indent: -40px; padding-left: 40px;">
Davis, S. and Meltzer, P. S. GEOquery: a bridge between the Gene Expression 
Omnibus (GEO) and BioConductor. Bioinformatics, 2007, 14, 1846-1847

Essayan-Perez, Sofia, and Thomas C Südhof. “Neuronal γ-Secretase Regulates 
Lipid Metabolism, Linking Cholesterol to Synaptic Dysfunction in Alzheimer’s 
Disease.” Neuron, vol. 111, no. 20, 1 Oct. 2023, pp. 3176-3194.e7, 
https://doi.org/10.1016/j.neuron.2023.07.005.

Robinson MD, McCarthy DJ and Smyth GK (2010). edgeR: a Bioconductor package for 
differential expression analysis of digital gene expression data. 
Bioinformatics 26, 139-140

Mapping identifiers for the integration of genomic datasets with the 
R/Bioconductor package biomaRt. Steffen Durinck, Paul T. Spellman, Ewan Birney 
and Wolfgang Huber, Nature Protocols 4, 1184-1191 (2009).

Martin, Fergal J., et al. “Ensembl 2023.” Nucleic Acids Research, 
vol. 51, no. D1, 6 Jan. 2023, pp. D933–D941, pubmed.ncbi.nlm.nih.gov/36318249/, 
https://doi.org/10.1093/nar/gkac958.
  
Wickham H, Henry L (2023). _purrr: Functional Programming Tools_. R package 
version 1.0.2, https://github.com/tidyverse/purrr, 
<https://purrr.tidyverse.org/>.

Wickham H, François R, Henry L, Müller K, Vaughan D (2023). _dplyr: A Grammar 
of Data Manipulation_. R package version 1.1.4, 
<https://CRAN.R-project.org/package=dplyr>.
  
Yihui Xie (2015) Dynamic Documents with R and knitr. 2nd edition. Chapman and 
Hall/CRC. ISBN 978-1498716963
</div>
